File: src/Binary_Tree/questions/subTree/SubTree.java
package Binary_Tree.questions.subTree;

// program to check whether a tree is a subTree of another Tree
public class SubTree {

    static class Node {
        int data;
        Node left, right;

        public Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }



    static class BinaryTree {

        static int index = -1;
        public static Node buildTree(int[] nodes) {

            index++;
            if(nodes[index] == -1)
                return null;

            Node newNode = new Node(nodes[index]);
            newNode.left = buildTree(nodes);
            newNode.right = buildTree(nodes);

            return newNode;

        }
    }

    public static boolean isIdentical(Node root, Node subRoot) {

        if(root == null && subRoot == null)
            return true;

        if(root == null || subRoot == null)
            return false;

        if(root.data == subRoot.data)
        return isIdentical(root.left, subRoot.left) && isIdentical(root.right, subRoot.right);

        return false;
    }

    public static boolean isSubTree(Node root, Node subRoot) {

        //case 1: if root of main tree is null
        if(root == null)
            return false;

        //case 2: if root of subTree is equal to null;
        if(subRoot == null)
            return true;

        //case 3: if root of main tree is equal to root of subTree
        if(root.data == subRoot.data) {
            if(isIdentical(root, subRoot)) {
                return true;
            }
        }
        return isSubTree(root.left, subRoot) || isSubTree(root.right, subRoot);

        }

    public static void main(String[] args) {

        int[] mainTree = {1,2,4,-1,-1,5,-1,-1,3,-1,-1};

        int[] subTree = {2,4,-1,-1,5,-1,-1};

        BinaryTree tree = new BinaryTree();
        Node mainRoot = tree.buildTree(mainTree);
        Node subRoot = tree.buildTree(subTree);

        System.out.println(isSubTree(mainRoot, subRoot));


    }
}

File: src/Binary_Tree/questions/subTree.py
class Node:
    # Node class to represent each node in the binary tree
    def __init__(self, val):
        self.data = val
        self.left = self.right = None

from collections import deque

def buildTree(nodes):
    # Builds a binary tree from a list of node values (level order)
    if not nodes and nodes[0] is None:
        return None

    root = Node(nodes[0])  # Create root node
    q = deque([root])      # Queue for level order traversal
    idx = 1

    while q and idx < len(nodes):
        curr = q.popleft()
        # Assign left child if value exists
        if idx < len(nodes) and nodes[idx]:
            curr.left = Node(nodes[idx])
            q.append(curr.left)
        idx += 1

        # Assign right child if value exists
        if idx < len(nodes) and nodes[idx]:
            curr.right = Node(nodes[idx])
            q.append(curr.right)
        idx += 1
    return root

def preorder(root):
    # Preorder traversal (root, left, right)
    if root is None:
        return None

    print(root.data, end=' ')
    preorder(root.left)
    preorder(root.right)

def isIdentical(node, subroot):
    # Checks if two trees are identical
    if node is None and subroot is None:
        return True
    if node is None or subroot is None or node.data != subroot.data:
        return False

    # Recursively check left and right subtrees
    return isIdentical(node.left, subroot.left) and isIdentical(node.right, subroot.right)

def subTree(root, subroot):
    # Checks if subroot is a subtree of root
    if root is None:
        return False

    # If current node matches subroot, check for identical trees
    if root.data == subroot.data:
        if isIdentical(root, subroot):
            return True

    # Recursively check left and right subtrees
    return subTree(root.left, subroot) or subTree(root.right, subroot)

# Example usage
tree1 = [1,2,3,4,5,None, 6]
root = buildTree(tree1)
print('Main Tree')
preorder(root)
print('\n')

tree2 = [2,4,5]
subroot = buildTree(tree2)
print('sub Tree')
preorder(subroot)
print('\n')

print('is tree2 sub tree of tree1: ')
print(subTree(root, subroot))
